File: ts/src/compose.ts
/** Compose system prompts from modules + emit manifests. */

import { existsSync, readFileSync } from "node:fs";
import path from "node:path";

import { CompositionError, ValidationError } from "./errors.js";
import { fileHash } from "./hashing.js";
import type {
  Manifest,
  ManifestModule,
  Module,
  SkeletonConfig,
} from "./models.js";
import {
  conflictToDict,
  manifestFromDict,
  manifestToDict,
  skeletonConfig,
} from "./models.js";
import { parseModule, parseModules } from "./parse.js";
import { DEFAULT_REGISTRY, type TypeRegistry } from "./registry.js";
import { renderPrompt } from "./render.js";
import {
  discoverLibraryTraitNames,
  resolveConflicts,
  validateModules,
} from "./validate.js";

export interface CompositionResult {
  promptXml: string;
  manifest: Manifest;
  manifestJson(indent?: number): string;
}

function toResult(promptXml: string, manifest: Manifest): CompositionResult {
  return {
    promptXml,
    manifest,
    manifestJson(indent = 2) {
      return JSON.stringify(manifestToDict(manifest), null, indent) + "\n";
    },
  };
}

function moduleToManifestEntry(m: Module): ManifestModule {
  return {
    path: m.path,
    type: m.type,
    name: m.name,
    hash: m.hash,
    source: m.source,
    source_hash: m.sourceHash,
    origin: m.origin,
    adaptation: m.adaptation,
    mode: m.type === "speech" ? m.mode : undefined,
  };
}

function buildManifest(
  modules: Module[],
  resolutions: ReturnType<typeof resolveConflicts>,
  warnings: string[],
  skeleton: Required<SkeletonConfig>,
  timestamp?: string,
): Manifest {
  const promptModules = modules.filter(
    (m) => !(m.type === "speech" && m.mode === "rewriter"),
  );
  const rewriters = modules.filter(
    (m) => m.type === "speech" && m.mode === "rewriter",
  );
  return {
    skeleton_version: skeleton.version,
    timestamp: timestamp ?? new Date().toISOString(),
    modules: promptModules.map(moduleToManifestEntry),
    conflict_rules: resolutions.map(conflictToDict),
    rewriter_stack: rewriters.map(moduleToManifestEntry),
    warnings,
  };
}

function isModule(x: string | Module): x is Module {
  return typeof x === "object" && x !== null && "type" in x && "hash" in x;
}

export interface ComposeOptions {
  skeleton?: SkeletonConfig;
  moduleRoot?: string;
  libraryRoot?: string;
  registry?: TypeRegistry;
  timestamp?: string;
}

export function compose(
  identity: string | Module,
  modules: Array<string | Module> = [],
  options: ComposeOptions = {},
): CompositionResult {
  const skeleton = skeletonConfig(options.skeleton);
  const registry = options.registry ?? DEFAULT_REGISTRY;

  const ensure = (item: string | Module): Module =>
    isModule(item)
      ? item
      : parseModule(item, {
          moduleRoot: options.moduleRoot,
          registry,
        });

  const identityMod = ensure(identity);
  if (identityMod.type !== "identity") {
    throw new ValidationError(
      `identity argument must be type=identity, got ${identityMod.type}`,
    );
  }

  const parsed: Module[] = [identityMod, ...modules.map(ensure)];

  const libRoot =
    options.libraryRoot !== undefined
      ? options.libraryRoot
      : options.moduleRoot;
  const libraryNames = discoverLibraryTraitNames(libRoot);
  const warnings = validateModules(parsed, {
    libraryTraitNames: libraryNames,
    registry,
  });
  const traits = parsed.filter((m) => m.type === "trait");
  const resolutions = resolveConflicts(traits);
  const promptXml = renderPrompt(parsed, resolutions, skeleton);
  const manifest = buildManifest(
    parsed,
    resolutions,
    warnings,
    skeleton,
    options.timestamp,
  );
  return toResult(promptXml, manifest);
}

export interface ComposeFromManifestOptions extends ComposeOptions {
  verifyHashes?: boolean;
}

export function composeFromManifest(
  manifestInput: Manifest | Record<string, unknown> | string,
  options: ComposeFromManifestOptions = {},
): CompositionResult {
  let manifest: Manifest;
  if (typeof manifestInput === "string") {
    manifest = manifestFromDict(
      JSON.parse(readFileSync(manifestInput, "utf-8")) as Record<string, unknown>,
    );
  } else if (
    "skeleton_version" in manifestInput &&
    "modules" in manifestInput &&
    Array.isArray((manifestInput as Manifest).modules)
  ) {
    // Already a Manifest-like or plain dict
    const m = manifestInput as Manifest;
    if (typeof m.timestamp === "string" && Array.isArray(m.conflict_rules)) {
      manifest =
        "rewriter_stack" in m && Array.isArray(m.rewriter_stack)
          ? m
          : manifestFromDict(manifestInput as Record<string, unknown>);
    } else {
      manifest = manifestFromDict(manifestInput as Record<string, unknown>);
    }
  } else {
    manifest = manifestFromDict(manifestInput as Record<string, unknown>);
  }

  let skeleton = skeletonConfig(
    options.skeleton ?? { version: manifest.skeleton_version },
  );
  if (skeleton.version !== manifest.skeleton_version) {
    skeleton = skeletonConfig({
      version: manifest.skeleton_version,
      output_rules: skeleton.output_rules,
    });
  }

  const allEntries = [...manifest.modules, ...manifest.rewriter_stack];
  if (!allEntries.length) {
    throw new ValidationError("manifest has no modules");
  }

  const verifyHashes = options.verifyHashes !== false;
  const paths: string[] = [];
  for (const entry of allEntries) {
    let filePath = entry.path;
    if (!existsSync(filePath)) {
      if (options.moduleRoot) {
        const alt = path.join(options.moduleRoot, entry.path);
        if (existsSync(alt)) filePath = alt;
        else {
          throw new CompositionError(
            `manifest module not found: ${entry.path}`,
          );
        }
      } else {
        throw new CompositionError(`manifest module not found: ${entry.path}`);
      }
    }
    if (verifyHashes) {
      const actual = fileHash(filePath);
      if (actual !== entry.hash) {
        throw new CompositionError(
          `hash mismatch for ${entry.path}: manifest=${entry.hash} actual=${actual}`,
        );
      }
      if (entry.source && entry.source_hash) {
        const root = options.moduleRoot ?? path.dirname(filePath);
        const sourcePath = path.resolve(root, entry.source);
        if (existsSync(sourcePath)) {
          const srcHash = fileHash(sourcePath);
          if (srcHash !== entry.source_hash) {
            throw new CompositionError(
              `source hash mismatch for ${entry.source}: ` +
                `manifest=${entry.source_hash} actual=${srcHash}`,
            );
          }
        }
      }
    }
    paths.push(filePath);
  }

  const parsed = parseModules(paths, {
    moduleRoot: options.moduleRoot,
    registry: options.registry,
  });
  const identities = parsed.filter((m) => m.type === "identity");
  if (!identities.length) {
    throw new ValidationError("manifest contains no identity module");
  }
  const identity = identities[0]!;
  const others = parsed.filter((m) => m !== identity);
  return compose(identity, others, {
    ...options,
    skeleton,
  });
}

File: src/persona_composer/compose.py
"""Compose system prompts from modules + emit manifests."""

from __future__ import annotations

import json
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from persona_composer.errors import CompositionError, ValidationError
from persona_composer.hashing import file_hash
from persona_composer.models import (
    Manifest,
    ManifestModule,
    Module,
    ModuleType,
    SkeletonConfig,
    SpeechMode,
)
from persona_composer.parse import parse_module, parse_modules
from persona_composer.registry import DEFAULT_REGISTRY, TypeRegistry
from persona_composer.render import render_prompt
from persona_composer.validate import (
    discover_library_trait_names,
    resolve_conflicts,
    validate_modules,
)


@dataclass
class CompositionResult:
    prompt_xml: str
    manifest: Manifest

    def manifest_json(self, *, indent: int = 2) -> str:
        return json.dumps(self.manifest.to_dict(), indent=indent) + "\n"


def _module_to_manifest_entry(m: Module) -> ManifestModule:
    return ManifestModule(
        path=str(m.path),
        type=m.type.value,
        name=m.name,
        hash=m.hash,
        source=m.source,
        source_hash=m.source_hash,
        origin=m.origin,
        adaptation=m.adaptation.value if m.adaptation else None,
        mode=m.mode.value if m.type == ModuleType.SPEECH else None,
    )


def _build_manifest(
    modules: list[Module],
    resolutions: list,
    warnings: list[str],
    skeleton: SkeletonConfig,
    *,
    timestamp: str | None = None,
) -> Manifest:
    prompt_modules = [
        m
        for m in modules
        if not (m.type == ModuleType.SPEECH and m.mode == SpeechMode.REWRITER)
    ]
    rewriters = [
        m
        for m in modules
        if m.type == ModuleType.SPEECH and m.mode == SpeechMode.REWRITER
    ]
    return Manifest(
        skeleton_version=skeleton.version,
        timestamp=timestamp or datetime.now(timezone.utc).isoformat(),
        modules=[_module_to_manifest_entry(m) for m in prompt_modules],
        conflict_rules=[r.to_dict() for r in resolutions],
        rewriter_stack=[_module_to_manifest_entry(m) for m in rewriters],
        warnings=warnings,
    )


def compose(
    identity: Path | Module,
    modules: list[Path | Module] | None = None,
    *,
    skeleton: SkeletonConfig | None = None,
    module_root: Path | None = None,
    library_root: Path | None = None,
    registry: TypeRegistry | None = None,
    timestamp: str | None = None,
) -> CompositionResult:
    """
    Compose a system prompt from an identity module plus optional modules.

    ``module_root`` resolves vendor ``source:`` paths.
    ``library_root`` is scanned for trait-name typo warnings (defaults to module_root).
    """
    skeleton = skeleton or SkeletonConfig()
    registry = registry or DEFAULT_REGISTRY
    modules = list(modules or [])

    parsed: list[Module] = []

    def _ensure(item: Path | Module) -> Module:
        if isinstance(item, Module):
            return item
        return parse_module(item, module_root=module_root, registry=registry)

    identity_mod = _ensure(identity)
    if identity_mod.type != ModuleType.IDENTITY:
        raise ValidationError(
            f"identity argument must be type=identity, got {identity_mod.type.value}"
        )
    parsed.append(identity_mod)
    for item in modules:
        parsed.append(_ensure(item))

    lib_root = library_root if library_root is not None else module_root
    library_names = discover_library_trait_names(lib_root)
    warnings = validate_modules(
        parsed, library_trait_names=library_names, registry=registry
    )
    traits = [m for m in parsed if m.type == ModuleType.TRAIT]
    resolutions = resolve_conflicts(traits)
    prompt_xml = render_prompt(parsed, resolutions, skeleton=skeleton)
    manifest = _build_manifest(
        parsed, resolutions, warnings, skeleton, timestamp=timestamp
    )
    return CompositionResult(prompt_xml=prompt_xml, manifest=manifest)


def compose_from_manifest(
    manifest: Manifest | dict[str, Any] | Path,
    *,
    skeleton: SkeletonConfig | None = None,
    module_root: Path | None = None,
    library_root: Path | None = None,
    registry: TypeRegistry | None = None,
    verify_hashes: bool = True,
    timestamp: str | None = None,
) -> CompositionResult:
    """Recompose from a saved manifest (recipe path)."""
    if isinstance(manifest, Path):
        data = json.loads(manifest.read_text(encoding="utf-8"))
        manifest = Manifest.from_dict(data)
    elif isinstance(manifest, dict):
        manifest = Manifest.from_dict(manifest)

    skeleton = skeleton or SkeletonConfig(version=manifest.skeleton_version)
    if skeleton.version != manifest.skeleton_version:
        # Allow override but default to manifest's version string on the SkeletonConfig
        skeleton = SkeletonConfig(
            version=manifest.skeleton_version,
            output_rules=skeleton.output_rules,
        )

    all_entries = list(manifest.modules) + list(manifest.rewriter_stack)
    if not all_entries:
        raise ValidationError("manifest has no modules")

    paths: list[Path] = []
    for entry in all_entries:
        path = Path(entry.path)
        if not path.is_file():
            # Try relative to module_root
            if module_root is not None:
                alt = module_root / entry.path
                if alt.is_file():
                    path = alt
                else:
                    raise CompositionError(f"manifest module not found: {entry.path}")
            else:
                raise CompositionError(f"manifest module not found: {entry.path}")
        if verify_hashes:
            actual = file_hash(path)
            if actual != entry.hash:
                raise CompositionError(
                    f"hash mismatch for {entry.path}: "
                    f"manifest={entry.hash} actual={actual}"
                )
            if entry.source and entry.source_hash:
                root = module_root or path.parent
                source_path = (root / entry.source).resolve()
                if source_path.is_file():
                    src_hash = file_hash(source_path)
                    if src_hash != entry.source_hash:
                        raise CompositionError(
                            f"source hash mismatch for {entry.source}: "
                            f"manifest={entry.source_hash} actual={src_hash}"
                        )
        paths.append(path)

    # Identify identity among parsed modules
    parsed = parse_modules(paths, module_root=module_root, registry=registry)
    identities = [m for m in parsed if m.type == ModuleType.IDENTITY]
    if not identities:
        raise ValidationError("manifest contains no identity module")
    identity = identities[0]
    others = [m for m in parsed if m is not identity]
    return compose(
        identity,
        others,
        skeleton=skeleton,
        module_root=module_root,
        library_root=library_root,
        registry=registry,
        timestamp=timestamp,
    )
